python/cvmfs/repository.py

```diff
@@ -198,38 +198,233 @@ def _pop_catalog(self):
+            return open(full_path, 'rb')
+        return None
+
+
+    def clear(self):
+        data_path = os.path.join(self._cache_dir, 'data')
+        shutil.rmtree(data_path)
+        self._create_cache_structure()
+
+
+
+class Fetcher:
+    """ Abstract wrapper around a Fetcher """
+
+    __metadata__ = abc.ABCMeta
+
+    def __init__(self, source, cache_dir=''):
+        self.cache = Cache(cache_dir)
+        self.source = source
+
+
+    def _make_file_uri(self, file_name):
+        return os.path.join(self.source, file_name)
+
+
+    @staticmethod
+    def _write_content_into_file(content, opened_file):
+        """ Writes content into the opened file. The file must have
+         write permission. The file is closed after after the operation
+        """
+        opened_file.write(content)
+        opened_file.close()
+
+
+    def retrieve_file(self, file_name):
+        """ Method to retrieve a file from the cache if exists, or from
+        the repository if it doesn't
+        """
+        cached_file = self.cache.get(file_name)
+        if not cached_file:
+            return self._retrieve_file(file_name)
+        return cached_file
+
+
+    @abc.abstractmethod
+    def _retrieve_file(self, file_name):
+        """ Abstract method to retrieve a file from the repository """
+        pass
+
+
+    @abc.abstractmethod
+    def retrieve_raw_file(self, file_name):
+        """ Abstract method to retrieve a raw file from the repository """
+        pass
+
+
+
+class LocalFetcher(Fetcher):
+    """ Retrieves files only from the local cache """
+
+    def __init__(self, local_repo, cache_dir=''):
+        Fetcher.__init__(self, local_repo, cache_dir)
+
+
+    def _retrieve_file(self, file_name):
+        full_path = self._make_file_uri(file_name)
+        if os.path.exists(full_path):
+            compressed_file = open(full_path, 'r')
+            decompressed_content = zlib.decompress(compressed_file.read())
+            compressed_file.close()
+            cached_file = self.cache.add(file_name)
+            self._write_content_into_file(decompressed_content, cached_file)
+            return self.cache.get(file_name)
+        else:
+            raise FileNotFoundInRepository(file_name)
+
+
+    def retrieve_raw_file(self, file_name):
+        """ Retrieves the file directly from the source """
+        full_path = self._make_file_uri(file_name)
+        if os.path.exists(full_path):
+            cached_file = self.cache.add(file_name)
+            compressed_file = open(full_path, 'r')
+            self._write_content_into_file(compressed_file.read(), cached_file)
+            compressed_file.seek(0)
+            return compressed_file
+        raise FileNotFoundInRepository(file_name)
+
+
+
+class RemoteFetcher(Fetcher):
+    """ Retrieves files from the local cache if found, and from
+    remote otherwise
+    """
+
+    def __init__(self, repo_url, cache_dir=''):
+        Fetcher.__init__(self, repo_url, cache_dir)
```

Comment:
So far it doesn't inherit from object, so I would have to change that. But yes, no problem with that